pkg/solidity-utils/contracts/helpers/SingletonAuthentication.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "@balancer-labs/v2-balancer-interfaces/contracts/solidity-utils/helpers/IAuthorizerAdaptor.sol";
+import "@balancer-labs/v2-balancer-interfaces/contracts/vault/IVault.sol";
+
+import "./Authentication.sol";
+
+abstract contract SingletonAuthentication is Authentication, IAuthorizerAdaptor {
+    IVault private immutable _vault;
+
+    // Use the contract's own address to disambiguate action identifiers
+    constructor(IVault vault) Authentication(bytes32(uint256(address(this)))) {
+        _vault = vault;
+    }
+
+    /**
+     * @notice Returns the Balancer Vault
+     */
+    function getVault() public view override returns (IVault) {
+        return _vault;
+    }
+
+    /**
+     * @notice Returns the Authorizer
+     */
+    function getAuthorizer() public view override returns (IAuthorizer) {
+        return getVault().getAuthorizer();
+    }
+
+    function _canPerform(bytes32 actionId, address account) internal view override returns (bool) {
+        return _canPerform(actionId, account, address(this));
+    }
+
+    function _canPerform(
+        bytes32 actionId,
+        address account,
+        address where
+    ) internal view returns (bool) {
+        return getAuthorizer().canPerform(actionId, account, where);
+    }
+
+    /**
+     * Concrete authorizer adapters must override - reverts if called at this level
+     */
+    function performAction(address, bytes calldata) external payable virtual override returns (bytes memory) {
+        _revert(Errors.INVALID_OPERATION);
+    }
```

Comment:
This does not belong here (nor in the interface) - `performAction` is strictly an AuthorizerAdaptor concept.